test/cpp/end2end/subchannel_metrics.cc

```diff
@@ -0,0 +1,206 @@
+//
+// Copyright 2025 gRPC authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#include <grpc/support/port_platform.h>
+#include <grpcpp/create_channel.h>
+#include <grpcpp/server_builder.h>
+
+#include <string>
+
+#include "src/core/telemetry/metrics.h"
+#include "src/proto/grpc/testing/echo.grpc.pb.h"
+#include "test/core/test_util/fake_stats_plugin.h"
+#include "test/core/test_util/port.h"
+#include "test/core/test_util/test_config.h"
+#include "test/core/test_util/tls_utils.h"
+#include "test/cpp/end2end/connection_attempt_injector.h"
+#include "gtest/gtest.h"
+
+namespace grpc {
+namespace testing {
+namespace {
+
+bool WaitForChannelState(
+    Channel* channel,
+    absl::AnyInvocable<bool(grpc_connectivity_state)> predicate,
+    bool try_to_connect = false, int timeout_seconds = 3) {
+  const gpr_timespec deadline =
+      grpc_timeout_seconds_to_deadline(timeout_seconds);
+  while (true) {
+    grpc_connectivity_state state = channel->GetState(try_to_connect);
+    if (predicate(state)) break;
+    if (!channel->WaitForStateChange(state, deadline)) return false;
+  }
+  return true;
+}
+
+std::shared_ptr<Channel> CreateChannelWithBackoff(const std::string& target,
+                                                  int backoff_ms) {
+  ChannelArguments args;
+  args.SetInt(GRPC_ARG_INITIAL_RECONNECT_BACKOFF_MS, backoff_ms);
+  args.SetInt(GRPC_ARG_MIN_RECONNECT_BACKOFF_MS, backoff_ms);
+  args.SetInt(GRPC_ARG_MAX_RECONNECT_BACKOFF_MS, backoff_ms);
+  return grpc::CreateCustomChannel("ipv4:" + target,
+                                   grpc::InsecureChannelCredentials(), args);
+}
+
+class MinimalEchoService : public grpc::testing::EchoTestService::Service {
+ public:
+  grpc::Status Echo(grpc::ServerContext* context,
+                    const grpc::testing::EchoRequest* request,
+                    grpc::testing::EchoResponse* response) override {
+    response->set_message(request->message());
+    return grpc::Status::OK;
+  }
+};
+
+class SubchannelMetricsTest : public ::testing::Test {
```

Comment:
Rather than buyilding a whole new test framework for this, it might be easier to just add these tests to the existing framework in client_lb_end2end_test.cc.  It already has a bunch of machinery for creating clients and servers, waiting for channel state changes, etc.